engine/internal/runtime/manager.go

```diff
@@ -0,0 +1,214 @@
+	delete(m.pendingRuntimes, modelID)
+}
+
+func (m *Manager) markRuntimeReady(modelID, address string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if r, ok := m.pendingRuntimes[modelID]; ok {
+		close(r)
+		delete(m.pendingRuntimes, modelID)
+		m.readyRuntimes[modelID] = runtime{address: address}
+	}
+}
+
+func (m *Manager) isReady(modelID string) bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	_, ok := m.readyRuntimes[modelID]
+	return ok
+}
+
+func (m *Manager) isPending(modelID string) bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	_, ok := m.pendingRuntimes[modelID]
+	return ok
+}
+
+// GetLLMAddress returns the address of the LLM.
+func (m *Manager) GetLLMAddress(modelID string) string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if r, ok := m.readyRuntimes[modelID]; ok {
+		return r.address
+	}
+	return ""
+}
+
+// ListSyncedModelIDs returns the list of models that are synced.
+func (m *Manager) ListSyncedModelIDs(ctx context.Context) []string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	var modelIDs []string
+	for id := range m.readyRuntimes {
+		modelIDs = append(modelIDs, id)
+	}
+	return modelIDs
+}
+
+// ListInProgressModels returns the list of models that are in progress.
+func (m *Manager) ListInProgressModels() []string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	var modelIDs []string
+	for id := range m.pendingRuntimes {
+		modelIDs = append(modelIDs, id)
+	}
+	return modelIDs
+}
+
+// PullModel pulls the model from the model manager.
+func (m *Manager) PullModel(ctx context.Context, modelID string) error {
+	if m.isReady(modelID) {
+		return nil
+	}
+
+	var done chan struct{}
+	m.mu.Lock()
+	if ch, ok := m.pendingRuntimes[modelID]; ok {
+		m.mu.Unlock()
+		done = ch
+
+	} else {
+		done = make(chan struct{})
+		m.pendingRuntimes[modelID] = done
+		m.mu.Unlock()
+
+		if err := m.rtClient.DeployRuntime(ctx, modelID); err != nil {
+			m.mu.Lock()
+			delete(m.pendingRuntimes, modelID)
+			m.mu.Unlock()
+			return err
+		}
+	}
+
+	ctrl.LoggerFrom(ctx).Info("Waiting for runtime to be ready", "model", modelID)
+	select {
+	case <-done:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return nil
+}
+
+// Reconcile reconciles the runtime.
+func (m *Manager) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	var sts appsv1.StatefulSet
+	if err := m.k8sClient.Get(ctx, req.NamespacedName, &sts); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	modelID := sts.GetAnnotations()[modelAnnotationKey]
+
+	if !sts.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&sts, finalizerKey) {
+			return ctrl.Result{}, nil
+		}
+		m.deleteRuntime(modelID)
+		controllerutil.RemoveFinalizer(&sts, finalizerKey)
+		if err := m.k8sClient.Update(ctx, &sts); err != nil {
+			log.Error(err, "Failed to remove finalizer")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if sts.Status.ReadyReplicas > 0 && m.isPending(modelID) {
```

Comment:
Slightly prefer removing `m.isPending`.
```suggestion
	if sts.Status.ReadyReplicas > 0 {
```

No concrete issue, but as `m.isPending` and `m.MarkRuntimeReady` acquire the lock separately, it requires more reasoning to guarantee correctness.

Reply:
I set this because I thought that, in many cases, it would be completed with a single read lock. This check is called when resources are changed or during periodic sync, but so far, markRuntimeReady has only been needed once.